pkg/cloudprovider/azure.go

```diff
@@ -1,37 +1,294 @@
 package cloudprovider
 
 import (
+	"context"
+	"fmt"
 	"net"
+	"strings"
 
+	network "github.com/Azure/azure-sdk-for-go/profiles/latest/network/mgmt/network"
+	compute "github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2020-06-30/compute"
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/adal"
+	azureapi "github.com/Azure/go-autorest/autorest/azure"
 	corev1 "k8s.io/api/core/v1"
+	utilnet "k8s.io/utils/net"
 )
 
 const (
-	azure = "azure"
+	azure                         = "azure"
+	defaultAzurePrivateIPCapacity = 256
 )
 
 // Azure implements the API wrapper for talking
 // to the Azure cloud API
 type Azure struct {
-	cloud CloudProvider
+	CloudProvider
+	resourceGroup        string
+	vmClient             compute.VirtualMachinesClient
+	virtualNetworkClient network.VirtualNetworksClient
+	networkClient        network.InterfacesClient
 }
 
 func (a *Azure) initCredentials() error {
+	clientID, err := a.readSecretData("azure_client_id")
+	if err != nil {
+		return err
+	}
+	tenantID, err := a.readSecretData("azure_tenant_id")
+	if err != nil {
+		return err
+	}
+	clientSecret, err := a.readSecretData("azure_client_secret")
+	if err != nil {
+		return err
+	}
+	subscriptionID, err := a.readSecretData("azure_subscription_id")
+	if err != nil {
+		return err
+	}
+	a.resourceGroup, err = a.readSecretData("azure_resourcegroup")
+	if err != nil {
+		return err
+	}
+	authorizer, err := a.getAuthorizer(clientID, clientSecret, tenantID)
+	if err != nil {
+		return err
+	}
+
+	a.vmClient = compute.NewVirtualMachinesClient(subscriptionID)
+	a.vmClient.Authorizer = authorizer
+	a.vmClient.AddToUserAgent(azure)
+
+	a.networkClient = network.NewInterfacesClient(subscriptionID)
+	a.networkClient.Authorizer = authorizer
+	a.networkClient.AddToUserAgent(azure)
+
+	a.virtualNetworkClient = network.NewVirtualNetworksClient(subscriptionID)
+	a.virtualNetworkClient.Authorizer = authorizer
+	a.virtualNetworkClient.AddToUserAgent(azure)
 	return nil
 }
 
 func (a *Azure) AssignPrivateIP(ip net.IP, node *corev1.Node) (interface{}, error) {
-	return nil, nil
+	instance, err := a.getInstance(node)
+	if err != nil {
+		return nil, err
+	}
+	networkInterface := network.Interface{}
+	for _, netif := range *instance.NetworkProfile.NetworkInterfaces {
+		if *netif.Primary {
```

Comment:
well, if we're assigning egress IPs to it then it must be a `Node` with status `Ready`, which definitely _strongly implies_ the existence of a NIC :slightly_smiling_face: .

Maybe just run through the code and for each deref, think about whether it could plausibly be `nil`, and make sure there are checks in those cases, but not in the "can't happen" cases.